board/safety/safety_volkswagen.h

```diff
@@ -1,142 +1,232 @@
+
+    // Update in-motion state by sampling front wheel speeds
+    // Signal: ESP_19.ESP_VL_Radgeschw_02 (front left)
+    // Signal: ESP_19.ESP_VR_Radgeschw_02 (front right)
+    if ((bus == 0) && (addr == MSG_ESP_19)) {
+      int wheel_speed_fl = GET_BYTE(to_push, 4) | (GET_BYTE(to_push, 5) << 8);
+      int wheel_speed_fr = GET_BYTE(to_push, 6) | (GET_BYTE(to_push, 7) << 8);
+      if ((wheel_speed_fl > 0) || (wheel_speed_fr > 0)) {
+        volkswagen_moving = true;
+      } else {
+        volkswagen_moving = false;
+      }
+    }
 
-      // every RT_INTERVAL set the new limits
-      uint32_t ts_elapsed = get_ts_elapsed(ts, volkswagen_ts_last);
-      if (ts_elapsed > VOLKSWAGEN_RT_INTERVAL) {
-        volkswagen_rt_torque_last = desired_torque;
-        volkswagen_ts_last = ts;
+    // Update driver input torque samples
+    // Signal: EPS_01.Driver_Strain (absolute torque)
+    // Signal: EPS_01.Driver_Strain_VZ (direction)
+    if ((bus == 0) && (addr == MSG_EPS_01)) {
+      int torque_driver_new = GET_BYTE(to_push, 5) | ((GET_BYTE(to_push, 6) & 0x1F) << 8);
+      int sign = (GET_BYTE(to_push, 6) & 0x80) >> 7;
+      if (sign == 1) {
+        torque_driver_new *= -1;
       }
+      update_sample(&volkswagen_torque_driver, torque_driver_new);
     }
 
-    // no torque if controls is not allowed
-    if (!controls_allowed && (desired_torque != 0)) {
-      violation = true;
+    // Update ACC status from drivetrain coordinator for controls-allowed state
+    // Signal: TSK_06.TSK_Status
+    if ((bus == 0) && (addr == MSG_TSK_06)) {
+      int acc_status = (GET_BYTE(to_push, 3) & 0x7);
+      controls_allowed = ((acc_status == 3) || (acc_status == 4) || (acc_status == 5)) ? 1 : 0;
     }
 
-    // reset to 0 if either controls is not allowed or there's a violation
-    if (violation || !controls_allowed) {
-      volkswagen_desired_torque_last = 0;
-      volkswagen_rt_torque_last = 0;
+    // Exit controls on rising edge of gas press
+    // Signal: Motor_20.MO_Fahrpedalrohwert_01
+    if ((bus == 0) && (addr == MSG_MOTOR_20)) {
+      int gas = (GET_BYTES_04(to_push) >> 12) & 0xFF;
+      if ((gas > 0) && (volkswagen_gas_prev == 0)) {
+        controls_allowed = 0;
+      }
+      volkswagen_gas_prev = gas;
+    }
+
+    // Exit controls on rising edge of brake press
+    // Signal: ESP_05.ESP_Fahrer_bremst
+    if ((bus == 0) && (addr == MSG_ESP_05)) {
+      bool brake_pressed = (GET_BYTE(to_push, 3) & 0x4) >> 2;
+      if (brake_pressed && (!(volkswagen_brake_pressed_prev) || volkswagen_moving)) {
+        controls_allowed = 0;
+      }
+      volkswagen_brake_pressed_prev = brake_pressed;
+    }
+
+    // If there are HCA messages on bus 0 not sent by OP, there's a relay problem
+    if ((safety_mode_cnt > RELAY_TRNS_TIMEOUT) && (bus == 0) && (addr == MSG_HCA_01)) {
+      relay_malfunction = true;
+    }
+  }
+  return valid;
+}
+
+static bool volkswagen_steering_check(int desired_torque) {
```

Comment:
Having abstracted this once for MQB vs PQ, it seems to me it should be abstracted globally for use by all ports...

Reply:
yeah, that would be a nice refactor. All cars except for honda have more or less these checks.